### ISSUE TITLE
Find a way to find the executable creation time that is not OS specific

### ISSUE THREAD

**Issue by rlepigre** [rlepigre/lambdapi#5]
https://github.com/rlepigre/lambdapi/blob/7a256a6916501a9e9b9d0a9eb93194c258c31e62/files.ml#L42

**Comment by fblanqui** [comment]
Isn't the name of the executable in Sys.argv.(0)?

**Comment by rlepigre** [comment]
No, only the command that was called. There is no way (that I know) of recovering the path to the actual binary from that.

**Comment by craff** [comment]
Salut 

The initial dir une Windows ;-)

a+
Christophe

Le 24 janvier 2018 11:26:14 GMT-03:00, Rodolphe Lepigre <notifications@github.com> a écrit :
>No, only the command that was called. There is no way (that I know) of
>recovering the path to the actual binary from that.
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/rlepigre/lambdapi/issues/5#issuecomment-360151723

-- 
Envoyé de mon appareil Android avec Courriel K-9 Mail. Veuillez excuser ma brièveté.

**Comment by fblanqui** [comment]
I do not understand what you want to do here. What is the "compiled 
program" the modification time you are looking for?


Le 24/01/2018 à 15:26, Rodolphe Lepigre a écrit :

**Comment by rlepigre** [comment]
I'm looking for the time at which the binary of the program was compiled. This information is useful if you want to check if a certain ".dko" was produced prior to the compilation of the program, which would mean that the ".dko" is outdated and needs to be produced again.

**Comment by craff** [comment]
On 18-01-24 14:26:14, Rodolphe Lepigre wrote:
> No, only the command that was called. There is no way (that I know) of
> recovering the path to the actual binary from that.

/proc/self/exe
is always a symbolic link that points to your running process.

raffalli@delli7:~/Programming/pml$ ls -l /proc/self/exe
lrwxrwxrwx 1 raffalli raffalli 0 janv. 24 17:51 /proc/self/exe -> /bin/ls

Là, c'est même drôle! J'attendais /bin/bash

A+
Christophe

**Comment by rlepigre** [comment]
Bah justement, c'est ce que je fais pour l'instant (/proc/self/exe), mais c'est visiblement pas une méthode qui marche sur tous les systèmes (même Unix-like).

Ce que tu obtiens pour "ls" c'est normal, vu que la chaîne "/proc/self/exe" est le contenu de argv[2], et que c'est "ls" qui va faire le "readdir" pour regarder vraiment le ficher... D'ailleurs t'as pas honte de faire ce genre d'expérience dans le répertoire de PML!

**Comment by fblanqui** [comment]
What is the difference between /proc/self/exe and argv.(0)?


Le 24/01/2018 à 22:55, Rodolphe Lepigre a écrit :

**Comment by rlepigre** [comment]
`"/proc/self/exe"` is the absolute path to a file, which is a symbolic link to the currently running binary, and `argv.(0)` is the command with which the program was called. For instance, if we did install `lambdapi.native` as `/usr/local/bin/lambdapi` then calling the command `lambdapi` would result in `argv.(0)` being exaclty `"lambdapi"`, which is not enough to know where the actual file is (`/usr/local/bin/lambdapi`).

**Comment by rlepigre** [comment]
Following Christophe's suggestion (by email), I changed the code a little bit to take advantage of the fact that marshaled closures can only be read by the binary that wrote them in the first place. In the same time I introduced a small change of behavior: if one attempts to load an incompatible object file, the program fails (instead of producing a new object file). This will be better if we ever install such files in a place where write permission is not given.
